crates/biome_deserialize/src/visitor.rs

```diff
@@ -1,39 +1,774 @@
+        self,
+        _items: impl Iterator<Item = SyntaxNode<L>>,
+        range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Value> {
+        self.expect(range, diagnostics);
+        None
+    }
+
+    fn expect(self, range: TextRange, diagnostics: &mut Vec<DeserializationDiagnostic>) {
+        diagnostics.push(DeserializationDiagnostic::new_incorrect_type(self, range));
+    }
+}
+
+pub trait Deserializable: Sized {
+    fn deserialize<L: DeserializableLanguage>(
+        value: SyntaxNode<L>,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self>;
+}
+
+pub trait DeserializableLanguage: Language {
+    fn deserialize<V: DeserializationVisitor>(
+        value: SyntaxNode<Self>,
+        visitor: V,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<V::Value>;
+}
+
+impl Deserializable for () {
+    fn deserialize<L: DeserializableLanguage>(
+        value: SyntaxNode<L>,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        diagnostics.push(
+            DeserializationDiagnostic::new("No value is allowed here.")
+                .with_range(value.text_trimmed_range()),
+        );
+        None
+    }
+}
+
+impl Deserializable for bool {
+    fn deserialize<L: DeserializableLanguage>(
+        value: SyntaxNode<L>,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        struct Visitor;
+        impl DeserializationVisitor for Visitor {
+            type Value = bool;
+            fn visit_bool(
+                self,
+                value: bool,
+                _range: TextRange,
+                _diagnostics: &mut Vec<DeserializationDiagnostic>,
+            ) -> Option<Self::Value> {
+                Some(value)
+            }
+        }
+        impl Display for Visitor {
+            fn fmt(&self, fmt: &mut Formatter) -> std::io::Result<()> {
+                write!(fmt, "boolean")
+            }
+        }
+        L::deserialize(value, Visitor, diagnostics)
+    }
+}
+
+impl Deserializable for f64 {
+    fn deserialize<L: DeserializableLanguage>(
+        value: SyntaxNode<L>,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self> {
+        L::deserialize(value, F64DeserializationVisitor, diagnostics)
+    }
+}
+
+struct F64DeserializationVisitor;
+impl DeserializationVisitor for F64DeserializationVisitor {
+    type Value = f64;
+
+    fn visit_f64(
+        self,
+        value: f64,
+        _range: TextRange,
         _diagnostics: &mut Vec<DeserializationDiagnostic>,
-    ) -> Option<()> {
-        unimplemented!("you should implement visit_map")
+    ) -> Option<Self::Value> {
+        Some(value)
+    }
+}
+
+impl Display for F64DeserializationVisitor {
+    fn fmt(&self, fmt: &mut Formatter) -> std::io::Result<()> {
+        write!(fmt, "float")
+    }
+}
+
+pub struct IntegerVisitor<T: TryFrom<i64>> {
```

Comment:
I forgot to add that it's usually idiomatic to add the boundaries only during the `impl` blocks.

We do this throughout the code base.


 I've seen it in a few places. The definition of the struct should be generic over `T`, and the implementation should define the boundary (which is like this already)